viewer/packages/tools/src/SmartOverlay/SmartOverlayTool.ts

```diff
@@ -0,0 +1,436 @@
+    this.clear();
+    this._viewer.domElement.removeEventListener('mousemove', this.onMouseMove);
+    this._viewer.off('click', this.onMouseClick);
+    this._events.disposed.fire();
+
+    this._events.hover.unsubscribeAll();
+    this._events.click.unsubscribeAll();
+
+    super.dispose();
+  }
+
+  private onMouseMove = (event: MouseEvent) => {
+    const { _textOverlay: textOverlay } = this;
+
+    const intersection = this.intersectPointsMarkers(event);
+
+    if (intersection) {
+      const { intersectedOverlay } = intersection;
+
+      const targetOverlay: SmartOverlay = {
+        text: intersectedOverlay.text,
+        id: intersectedOverlay.id,
+        position: intersectedOverlay.position,
+        infoOverlay: textOverlay
+      };
+
+      this.positionTextOverlay(event);
+      textOverlay.innerText = targetOverlay.text;
+
+      this._events.hover.fire({ targetOverlay: targetOverlay, mousePosition: event });
+    } else {
+      textOverlay.style.opacity = '0';
+    }
+  }
+
+  private onMouseClick = (event: PointerEventData) => {
+    const intersection = this.intersectPointsMarkers({ clientX: event.offsetX, clientY: event.offsetY });
+
+    if (intersection) {
+      const { intersectedOverlay } = intersection;
+
+      const targetOverlay: SmartOverlay = {
+        text: intersectedOverlay.text,
+        id: intersectedOverlay.id,
+        position: intersectedOverlay.position,
+        infoOverlay: this._textOverlay
+      };
+
+      this._events.click.fire({
+        targetOverlay: targetOverlay,
+        mousePosition: { clientX: event.offsetX, clientY: event.offsetY }
+      });
+    }
+  }
+
+  private positionTextOverlay(event: MouseEvent): void {
+    const { _textOverlay, _textOverlayVisible } = this;
+    _textOverlay.style.left = `${event.offsetX}px`;
+    _textOverlay.style.top = `${event.offsetY}px`;
+    _textOverlay.style.opacity = _textOverlayVisible ? '1' : '0';
+  }
+
+  private intersectPointsMarkers(mouseCoords: {
+    clientX: number;
+    clientY: number;
+  }): { intersectedIcon: Overlay3DIcon; intersectedOverlay: OverlayInfo } | null {
+    const { _viewer, _raycaster, _temporaryVec } = this;
+
+    const { x, y } = this.convertPixelCoordinatesToNormalized(mouseCoords, _viewer.domElement);
+    const camera = _viewer.cameraManager.getCamera();
+    const cameraDirection = camera.getWorldDirection(new THREE.Vector3());
+    _raycaster.setFromCamera(_temporaryVec.set(x, y), camera);
+
+    let intersection: [Overlay3DIcon, THREE.Vector3][] = [];
+
+    for (const points of this._overlayPoints) {
+      for (const icon of points.icons) {
+        if (icon.visible) {
+          const inter = icon.intersect(_raycaster.ray);
+          if (inter) {
+            intersection.push([icon, inter]);
+            icon.updateAdaptiveScale({
+              camera,
+              renderSize: _viewer.renderParameters.renderSize,
+              domElement: _viewer.canvas
+            });
+          }
+        }
```

Comment:
Yeah, after talking with Lars we decided that overlays should behave in the same way as 360 images, i.e. they will be visible event when occluded by geometry, so it will then make sense to always intersect them.